vmm/src/landlock.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright © 2024 Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use landlock::{
+    path_beneath_rules, Access, AccessFs, BitFlags, PathFdError, Ruleset, RulesetAttr,
+    RulesetCreated, RulesetCreatedAttr, RulesetError, ABI,
+};
+use std::io::Error as IoError;
+use std::path::PathBuf;
+use thiserror::Error;
+
+use crate::vm_config::LandlockConfig;
+
+pub const READ: u8 = 1 << 0;
+pub const WRITE: u8 = 1 << 1;
+pub const EXECUTE: u8 = 1 << 2;
+
+#[derive(Debug, Error)]
+pub enum LandlockError {
+    /// Failed to create Ruleset
+    #[error("Error creating ruleset: {0}")]
+    CreateRuleSet(#[source] RulesetError),
+
+    /// Failed to add rule
+    #[error("Error adding rule to ruleset: {0}")]
+    AddRule(#[source] RulesetError),
+
+    /// Failed to restrict self
+    #[error("Error restricting self: {0}")]
+    RestrictSelf(#[source] RulesetError),
+
+    /// Ruleset is empty
+    #[error("Ruleset was previously consumed, cannot add rules to it")]
+    RulesetConsumed,
+
+    /// Error opening Path
+    #[error("Error opening Path: {0}")]
+    OpenPath(#[source] PathFdError),
+
+    /// Error opening Device Path
+    #[error("Error opening Device Path: {0}")]
+    OpenDevicePath(#[source] IoError),
+}
+
+pub struct Landlock {
+    ruleset: Option<RulesetCreated>,
+    abi: ABI,
+}
+
+impl Landlock {
+    pub fn new() -> Result<Landlock, LandlockError> {
+        let abi = ABI::V3;
+
+        let file_access = AccessFs::from_all(abi);
+        let def_ruleset = Ruleset::default()
+            .handle_access(file_access)
+            .map_err(LandlockError::CreateRuleSet)?;
+
+        let ruleset = def_ruleset.create().map_err(LandlockError::CreateRuleSet)?;
+
+        Ok(Landlock {
+            ruleset: Some(ruleset),
+            abi,
+        })
+    }
+
+    pub fn add_rule(
+        &mut self,
+        path: PathBuf,
+        access: BitFlags<AccessFs>,
+    ) -> Result<(), LandlockError> {
+        // path_beneath_rules in landlock crate handles file and directory access rules.
+        // Incoming path/s are passed to path_beneath_rules, so that we don't
+        // have to worry about the type of the path.
+        let paths = vec![path.clone()];
+        let path_beneath_rules = path_beneath_rules(paths, access);
+        if let Some(ruleset) = &mut self.ruleset {
```

Comment:
Oh, I now understand your use of `RulesetConsumed`. Why do you need `self.ruleset` to be an `Option`? The library was designed to avoid as much as possible run time checks such as this one. This is why some methods consume the `self` object instead of passing it as a reference, and also why there is the `Compatible` trait which should handle all cases when Landlock is not supported.

Reply:
I made `self.ruleset` an `Option` because the ruleset is consumed whenever `add_rule` is called. I see `add_rule` returns the ruleset after consuming it. I can drop `Option` for self.ruleset and update it after every `add_rule`.

Reply:
After investigating this a bit, I realized `add_rules` consumes the passed `ruleset` and returns a new one. I updated the code to handle this properly.

This in turn requires me to consume `landlock` object and return a new one from the methods defined in my landlock.rs. This look a bit odd. Any suggestions on how to better handle this?

Reply:
Using a mutable reference should work because [`RulesetCreatedAttr` is implemented for `&mut RulesetCreated` too](https://landlock.io/rust-landlock/landlock/trait.RulesetCreatedAttr.html#impl-RulesetCreatedAttr-for-%26mut+RulesetCreated).

Reply:
updated